gcc/rust/util/rust-token-converter.cc

```diff
@@ -0,0 +1,586 @@
+{
+  std::vector<ProcMacro::TokenStream> trees;
+  trees.push_back (ProcMacro::TokenStream::make_tokenstream ());
+  for (auto &token : tokens)
+    {
+      switch (token->get_id ())
+	{
+	// Literals
+	case FLOAT_LITERAL:
+	  dispatch_float_literals (trees.back (), token);
+	  break;
+	case INT_LITERAL:
+	  dispatch_integer_literals (trees.back (), token);
+	  break;
+	// FIXME: Why does BYTE_CHAR_LITERAL is not handled by rustc ?
+	case CHAR_LITERAL: // TODO: UTF-8 handling
+	  trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+	    ProcMacro::Literal::make_char (token->as_string ()[0])));
+	  break;
+	case STRING_LITERAL:
+	  trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+	    ProcMacro::Literal::make_string (token->as_string ())));
+	  break;
+	  case BYTE_STRING_LITERAL: {
+	    auto str = token->as_string ();
+	    std::vector<uint8_t> data (str.begin (), str.end ());
+	    trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+	      ProcMacro::Literal::make_byte_string (data)));
+	  }
+	  break;
+	// Ident
+	case IDENTIFIER:
+	case ABSTRACT:
+	case AS:
+	case ASYNC:
+	case AUTO:
+	case BECOME:
+	case BOX:
+	case BREAK:
+	case CONST:
+	case CONTINUE:
+	case CRATE:
+	case DO:
+	case DYN:
+	case ELSE:
+	case ENUM_TOK:
+	case EXTERN_TOK:
+	case FINAL_TOK:
+	case FN_TOK:
+	case FOR:
+	case IF:
+	case IMPL:
+	case IN:
+	case LET:
+	case LOOP:
+	case MACRO:
+	case MATCH_TOK:
+	case MOD:
+	case MOVE:
+	case MUT:
+	case OVERRIDE_TOK:
+	case PRIV:
+	case PUB:
+	case REF:
+	case RETURN_TOK:
+	case SELF_ALIAS:
+	case SELF:
+	case STATIC_TOK:
+	case STRUCT_TOK:
+	case SUPER:
+	case TRAIT:
+	case TRY:
+	case TYPE:
+	case TYPEOF:
+	case UNSAFE:
+	case UNSIZED:
+	case USE:
+	case VIRTUAL:
+	case WHERE:
+	case WHILE:
+	case YIELD:
+	// Underscore is not a Punct, considered as an Ident
+	case UNDERSCORE:
+	// True and false are idents, not literals
+	// (https://doc.rust-lang.org/proc_macro/struct.Literal.html)
+	case FALSE_LITERAL:
+	case TRUE_LITERAL:
+	  trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+	    ProcMacro::Ident::make_ident (token->as_string ())));
+	  break;
+	// Joint punct
+	case OR:
+	case PIPE_EQ:
+	case CARET_EQ:
+	case RIGHT_SHIFT_EQ:
+	case RIGHT_SHIFT:
+	case GREATER_OR_EQUAL:
+	case MATCH_ARROW:
+	case LESS_OR_EQUAL:
+	case LEFT_SHIFT_EQ:
+	case LEFT_SHIFT:
+	case DIV_EQ:
+	case ELLIPSIS:
+	case DOT_DOT_EQ:
+	case DOT_DOT:
+	case RETURN_TYPE:
+	case MINUS_EQ:
+	case PLUS_EQ:
+	case ASTERISK_EQ:
+	case LOGICAL_AND:
+	case AMP_EQ:
+	case PERCENT_EQ:
+	case SCOPE_RESOLUTION:
+	case NOT_EQUAL:
+	  case EQUAL_EQUAL: {
+	    auto str = token->as_string ();
+	    auto it = str.cbegin ();
+	    for (; it != str.cend () - 1; it++)
+	      {
+		trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+		  ProcMacro::Punct::make_punct (*it, ProcMacro::JOINT)));
+	      }
+	    trees.back ().push (ProcMacro::TokenTree::make_tokentree (
+	      ProcMacro::Punct::make_punct (*it, ProcMacro::ALONE)));
```

Comment:
so we push back a `JOINT` punc token and an `ALONE` one?

Reply:
No, we push all characters but the last one as joint if any, and then the last one `alone`. This way both single and complex punctuation are handled.